Foundation/URLSession/NativeProtocol.swift

```diff
@@ -0,0 +1,754 @@
+        case .transferInProgress:                  return true
+        case .transferCompleted:                   return false
+        case .transferFailed:                      return false
+        case .waitingForRedirectCompletionHandler: return false
+        case .waitingForResponseCompletionHandler: return true
+        case .taskCompleted:                       return false
+        }
+    }
+    var isEasyHandlePaused: Bool {
+        switch self {
+        case .initial:                             return false
+        case .transferReady:                       return false
+        case .transferInProgress:                  return false
+        case .transferCompleted:                   return false
+        case .transferFailed:                      return false
+        case .waitingForRedirectCompletionHandler: return false
+        case .waitingForResponseCompletionHandler: return true
+        case .taskCompleted:                       return false
+        }
+    }
+}
+
+extension _NativeProtocol {
+    
+    enum _DataDrain {
+        /// Concatenate in-memory
+        case inMemory(NSMutableData?)
+        /// Write to file
+        case toFile(URL, FileHandle?)
+        /// Do nothing. Might be forwarded to delegate
+        case ignore
+    }
+    enum _Error: Error {
+        case parseSingleLineError
+        case parseCompleteHeaderError
+    }
+    
+    func errorCode(fileSystemError error: Error) -> Int {
+        func fromCocoaErrorCode(_ code: Int) -> Int {
+            switch code {
+            case CocoaError.fileReadNoSuchFile.rawValue:
+                return NSURLErrorFileDoesNotExist
+            case CocoaError.fileReadNoPermission.rawValue:
+                return NSURLErrorNoPermissionsToReadFile
+            default:
+                return NSURLErrorUnknown
+            }
+        }
+        switch error {
+        case let e as NSError where e.domain == NSCocoaErrorDomain:
+            return fromCocoaErrorCode(e.code)
+        default:
+            return NSURLErrorUnknown
+        }
+    }
+}
+
+extension _NativeProtocol._TransferState {
+    /// Transfer state that can receive body data, but will not send body data.
+    init(url: URL, bodyDataDrain: _NativeProtocol._DataDrain) {
+        self.url = url
+        self.bodyDataDrain = bodyDataDrain
+        self.response = nil
+        self.parsedResponseHeader = _NativeProtocol._ParsedResponseHeader()
+        self.requestBodySource = nil
+    }
+    
+    /// Transfer state that sends body data and can receive body data.
+    init(url: URL, bodyDataDrain: _NativeProtocol._DataDrain, bodySource: _BodySource) {
+        self.url = url
+        self.parsedResponseHeader = _NativeProtocol._ParsedResponseHeader()
+        self.response = nil
+        self.requestBodySource = bodySource
+        self.bodyDataDrain = bodyDataDrain
+    }
+    
+}
+
+struct _Delimiters {
+    /// *Carriage Return* symbol
+    static let CR: UInt8 = 0x0d
```

Comment:
Why are these `UInt8` and the others are `UnicodeScalar`? Worth a comment to indicate what the issue is.